### PR TITLE
SolverGMRES: Make loops in Arnoldi orthogonalization prefetcher-friendly 

### PR DESCRIPTION
**Pull request by kronbichler**
The orthogonalization done within `SolverGMRES` and `SolverFGMRES` for the deal.II vectors would previously lead to data access pattern that are unfriendly to data prefetchers on modern CPUs. I observed this because the orthogonalization was slower with the classical Gram-Schmidt than with the modified Gram-Schmidt algorithm, whereas I had expected the opposite.

This can be addressed by implementing a suitable loop blocking, i.e., we do run try to switch between all vectors in the inner loop, but only up to a small number, and then perform an outer loop. There should be no change in functionality.

### CODE REVIEW

File: source/lac/solver_gmres.cc [hunk 1]
```diff
@@ -53,61 +53,74 @@ namespace internal
...
-              for (unsigned int i = 0; i < n_vectors - 1; ++i)
+          const unsigned int loop_length_c =
+            locally_owned_size / n_lanes / inner_batch_size;
+          for (unsigned int c_block = 0; c_block < (loop_length_c + 63) / 64;
+               ++c_block)
+            for (unsigned int i_block = 0; i_block < (n_vectors + 7) / 8;
+                 ++i_block)
+              for (c = c_block * 64, j = c * n_lanes * inner_batch_size;
+                   c < std::min(loop_length_c, (c_block + 1) * 64);
+                   ++c, j += n_lanes * inner_batch_size)
```
Comment (bangerth) [inline source/lac/solver_gmres.cc:64]: Would you mind prefixing all of this with a description of why you break it into three loops? And perhaps include the original one loop into the comments?

Comment (kronbichler) [reply source/lac/solver_gmres.cc:64]: Excellent idea, I will do soon!

Comment (kronbichler) [reply source/lac/solver_gmres.cc]: I added a long comment in one function, and a shorter one with reference to the first comment in the second place I adjusted. Thanks for the discussion, I think this helps quite a bit.

**Review by bangerth — COMMENTED** [review]
Thank you!

**Review by bangerth — APPROVED** [review]

**Pull request merged**